drivers/network/dd/dc21x4/dc21x4.h

```diff
@@ -0,0 +1,536 @@
+
+NDIS_STATUS
+DcSetPower(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ NDIS_DEVICE_POWER_STATE PowerState);
+
+NDIS_STATUS
+DcAddWakeUpPattern(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ PNDIS_PM_PACKET_PATTERN PmPattern);
+
+NDIS_STATUS
+DcRemoveWakeUpPattern(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ PNDIS_PM_PACKET_PATTERN PmPattern);
+
+CODE_SEG("PAGE")
+VOID
+DcFreeAdapter(
+    _In_ __drv_freesMem(Mem) PDC21X4_ADAPTER Adapter);
+
+CODE_SEG("PAGE")
+VOID
+MediaInitializeMediaList(
+    _In_ PDC21X4_ADAPTER Adapter);
+
+CODE_SEG("PAGE")
+VOID
+NTAPI
+DcResetWorker(
+    _In_ PNDIS_WORK_ITEM WorkItem,
+    _In_opt_ PVOID Context);
+
+CODE_SEG("PAGE")
+DECLSPEC_NOINLINE
+VOID
+DcStopAdapter(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ BOOLEAN IsPowerTransitioning);
+
+VOID
+DcStartAdapter(
+    _In_ PDC21X4_ADAPTER Adapter);
+
+CODE_SEG("PAGE")
+NDIS_STATUS
+DcSoftReset(
+    _In_ PDC21X4_ADAPTER Adapter);
+
+CODE_SEG("PAGE")
+NDIS_STATUS
+DcReadEeprom(
+    _In_ PDC21X4_ADAPTER Adapter);
+
+CODE_SEG("PAGE")
+VOID
+DcInitTxRing(
+    _In_ PDC21X4_ADAPTER Adapter);
+
+CODE_SEG("PAGE")
+VOID
+DcInitRxRing(
+    _In_ PDC21X4_ADAPTER Adapter);
+
+ULONG
+DcEthernetCrc(
+    _In_reads_bytes_(Size) const VOID* Buffer,
+    _In_ ULONG Size);
+
+VOID
+DcDisableHw(
+    _In_ PDC21X4_ADAPTER Adapter);
+
+VOID
+DcStopTxRxProcess(
+    _In_ PDC21X4_ADAPTER Adapter);
+
+VOID
+DcWriteGpio(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ ULONG Value);
+
+VOID
+DcWriteSia(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ ULONG Csr13,
+    _In_ ULONG Csr14,
+    _In_ ULONG Csr15);
+
+BOOLEAN
+DcTestPacket(
+    _In_ PDC21X4_ADAPTER Adapter);
+
+CODE_SEG("PAGE")
+VOID
+DcSetupFrameInitialize(
+    _In_ PDC21X4_ADAPTER Adapter);
+
+BOOLEAN
+DcSetupFrameDownload(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ BOOLEAN WaitForCompletion);
+
+NDIS_STATUS
+DcApplyPacketFilter(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ ULONG PacketFilter);
+
+NDIS_STATUS
+DcUpdateMulticastList(
+    _In_ PDC21X4_ADAPTER Adapter);
+
+VOID
+DcPowerSave(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ BOOLEAN Enable);
+
+CODE_SEG("PAGE")
+BOOLEAN
+DcFindMiiPhy(
+    _In_ PDC21X4_ADAPTER Adapter);
+
+BOOLEAN
+MiiWrite(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ ULONG PhyAddress,
+    _In_ ULONG RegAddress,
+    _In_ ULONG Data);
+
+BOOLEAN
+MiiRead(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ ULONG PhyAddress,
+    _In_ ULONG RegAddress,
+    _Out_ PULONG Data);
+
+VOID
+HpnaWrite(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ ULONG RegAddress,
+    _In_ ULONG Data);
+
+VOID
+HpnaRead(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ ULONG RegAddress,
+    _Out_ PULONG Data);
+
+#if defined(__GNUC__)
+#define MEMCMP_HACK
+#endif
+
+#if defined(MEMCMP_HACK)
+BOOLEAN
+DcEqualMemory(
+    _In_reads_bytes_(Length) const VOID* Destination,
+    _In_reads_bytes_(Length) const VOID* Source,
+    _In_ SIZE_T Length);
+#undef NdisEqualMemory
+#define NdisEqualMemory DcEqualMemory
+#endif
```

Comment:
Why this hack is needed, does `NdisEqualMemory` function has some problem in GCC builds? (and I've just noticed a similar workaround in nvnet)

Reply:
Without this hack, it will not compile in debug mode.

Reply:
What's the error you obtain when you compile in debug mode?

Reply:
```
undefined reference to `memcmp'
collect2.exe: error: ld returned 1 exit status
```

Reply:
Added a new library to the list.